S3TransferUtilitySample/src/com/amazonaws/demo/s3transferutility/UploadActivity.java

```diff
@@ -519,85 +517,68 @@ private void beginUploadInBackground(String filePath) {
-        String[] selectionArgs = null;
-        // Uri is different in versions after KITKAT (Android 4.4), we need to
-        // deal with different Uris.
-        if (needToCheckUri && DocumentsContract.isDocumentUri(getApplicationContext(), uri)) {
-            if (isExternalStorageDocument(uri)) {
-                final String docId = DocumentsContract.getDocumentId(uri);
-                final String[] split = docId.split(":");
-                return Environment.getExternalStorageDirectory() + "/" + split[1];
-            } else if (isDownloadsDocument(uri)) {
-                final String id = DocumentsContract.getDocumentId(uri);
-                uri = ContentUris.withAppendedId(
-                        Uri.parse("content://downloads/public_downloads"), Long.valueOf(id));
-            } else if (isMediaDocument(uri)) {
-                final String docId = DocumentsContract.getDocumentId(uri);
-                final String[] split = docId.split(":");
-                final String type = split[0];
-                if ("image".equals(type)) {
-                    uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
-                } else if ("video".equals(type)) {
-                    uri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI;
-                } else if ("audio".equals(type)) {
-                    uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
+    private File getFile(Uri uri) throws FileNotFoundException {
+        final InputStream in = getContentResolver().openInputStream(uri);
+        OutputStream out = null;
+        try {
+            final File file = new File(getCacheDir(), getFileName(uri));
+            out = new FileOutputStream(file, false);
+            byte[] buffer = new byte[1024];
+            for (int len; (len = in.read(buffer)) != -1; ) {
+                out.write(buffer, 0, len);
+            }
+            return file;
+        } catch (IOException e) {
+            e.printStackTrace();
+        } finally {
+            try {
+                if (in != null) {
+                    in.close();
                 }
-                selection = "_id=?";
-                selectionArgs = new String[] {
-                        split[1]
-                };
+                if(out != null) {
+                    out.close();
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
             }
         }
+        return null;
+    }
+
+    private String getFileName(Uri uri) {
         if ("content".equalsIgnoreCase(uri.getScheme())) {
-            String[] projection = {
-                    MediaStore.Images.Media.DATA
+            final String[] projection = {
+                    MediaStore.Images.Media.DISPLAY_NAME
             };
-            Cursor cursor = null;
+            final String docId = DocumentsContract.getDocumentId(uri);
+            String selection = null;
+            String[] selectionArgs = null;
+            if(docId.contains(":")) {
```

Comment:
Can we add a `// the format of the thing being parsed here is expected to be <foo>` ?

Reply:
This is an artifact of the previous implementation, but I'm realizing it's actually not necessary.  Specifying selection and selectionArgs would be necessary if the `Uri` pointed to a directory, but since it already points to `Document`, we don't need to specific the documentId as a selectionArg.